src/components/menu/menu.ts

```diff
@@ -100,23 +116,65 @@ export class Menu extends Component<Props, SpreadsheetChildEnv> {
   static defaultProps = {
     depth: 1,
   };
+  private clickableMenuItems: Action[] = [];
   private subMenu: MenuState = useState({
     isOpen: false,
     position: null,
     scrollOffset: 0,
     menuItems: [],
+    menuIndex: this.props.shouldSelectFirstItem ? 0 : undefined,
+    shouldSelectFirstItem: false,
   });
   private menuRef = useRef("menu");
   private position: DOMCoordinates = useAbsoluteBoundingRect(this.menuRef);
+  private menuItemRefs: { [id: string]: { action: Action; ref: Ref<HTMLElement> } } = {};
+  // @ts-ignore: childrenHaveIcon will be used in XML
+  childrenHaveIcon: boolean = false;
 
   setup() {
+    this.props.menuItems.forEach((menuItem) => {
+      this.menuItemRefs[menuItem.id] = {
+        action: menuItem,
+        ref: useRef(menuItem.id),
+      };
+    });
     useExternalListener(window, "click", this.onExternalClick, { capture: true });
     useExternalListener(window, "contextmenu", this.onExternalClick, { capture: true });
+    onWillRender(() => {
```

Comment:
@hokolomopo Items in `this.clickableMenuItems` can change as you and Remi mentioned, and for now I didn't find a way to "monitor" it, because the state of menu items depends on env as well. 
I also tried to make `clickableMenuItems` as getter, but it has performance issue when there are too many menu items (e.g. insert->function->all). 
We discussed about `onWillRender` this afternoon and it's indeed not good. After looking into the menus on runbot master I didn't find any menu item that would become disabled (but still visible) when corporating (something like undo & redo is local). I only found freeze & unfreeze panel menu item can become invisible and visible from time to time when corporating. Hence I use `useEffect` together with the length of `menuItemsAndSeparators` to kind of "monitor" changes of `clickableMenuItems`. This is called much less times than `onWillRender`. And also in `useEffect` I used a simpler fallback (I simply reset the item selection 😅). 
Feel free to think about it twice when you're reviewing 😊